### ISSUE TITLE
Expanding does not work with clickToSelect

### ISSUE THREAD

**Issue by ZhuQinglei** [AllenFang/react-bootstrap-table#1063]
Hi, @AllenFang  I was still struggling with the managing expanding with row-select. I found the same issue here: [issue 1038](https://github.com/AllenFang/react-bootstrap-table/issues/1038) 

But unlike the solution provided(disable clicktoselect), I would like both clickToSelect = true and clickToExpand = true, because I need clickToSelect to hight the row selected, and clickToExpand to expand the row at the same time.

I wanted to:
- select and expand the row on row click // working 
- unselect and collapse the expandComponent on button click inside the expandComponent // not working

Currently the options{expanding: ...} is not working though I followed the example here: [manage-expanding](https://github.com/AllenFang/react-bootstrap-table/blob/master/examples/js/expandRow/manage-expanding.js)
 
I am managing the selection and expanding from the two redux function : 
- this.props.selectOrganization(row._id)  
- this.props.unSelectOrganization(row._id);

Here is the code for the redux function: 
`
export default function reducer(state = defaultState, action) {
    switch (action.type) {

        case "SELECT_ORGANIZATION": {
            console.log([...state.batchEdit.selectedOrganizations, action.org]);
            return {
                ...state,
                batchEdit: {
                    ...state.batchEdit,
                    selectedOrganizations: [...state.batchEdit.selectedOrganizations, action.org]
                }
            }
        }

        case "UNSELECT_ORGANIZATION": {
            // console.log(state.batchEdit.selectedOrganizations.filter(it => it !== action.org));
            return {
                ...state,
                batchEdit: {
                    ...state.batchEdit,
                    selectedOrganizations: state.batchEdit.selectedOrganizations.filter(it => it !== action.org)
                }
            }
        }
...
`

Here is my code for the table:
`
class OrganizationsTable extends Component {

    constructor(props) {
        super(props);
    }

    handleRowSelect(row, isSelected, e){
        if (isSelected) { 
            this.props.selectOrganization(row._id);
        } else {
            // delete from selectedRow array if unselected
            this.props.unSelectOrganization(row._id);
        }
        return false;
    }

    render() {

        const options = {
            expandRowBgColor: '#D9D9D9', // grey background
            sizePerPageList: [10, 15, 20, 50, 100], // Dropdown Options for rows per page
            sizePerPage: 10,
            expanding: this.props.selectedOrganizations, // not working as expected
        };

        const selectRow = {
            mode: 'checkbox', // multi-select
            className: 'clicked-row',
            clickToSelect: true,  // click to select, default is false
            clickToExpand: true,  // click to expand row, default is false
            selected: this.props.selectedOrganizations, // manage selection
            onSelect: this.handleRowSelect.bind(this),
        };

        var validateContext = Organizations.schema.namedContext("createNewOrganization");
        return (
            <div>
                <MyImporter importFunction='organizations.import'
                            message='You have successfully added all organizations' context={validateContext}
                            collections='Organizations'/>
                <BootstrapTable
                    data={ this.props.organizations }
                    options={ options } selectRow={ selectRow }
                    expandableRow={ () => {
                        return true // always return true as all rows are expandable
                    } }
                    expandComponent={(row) => {
                        return (<OrganizationDetail organization={ row }/>)
                    }}
                    exportCSV pagination>
                    <TableHeaderColumn dataField='_id' isKey hidden>ID</TableHeaderColumn>
                    <TableHeaderColumn dataField='organization_name'
                                       dataSort={ true }>Organization</TableHeaderColumn>
                    <TableHeaderColumn dataField='address' dataSort={ true }>Address</TableHeaderColumn>
                    <TableHeaderColumn dataField='post_code' dataSort={ true }>Postal Code</TableHeaderColumn>
                    <TableHeaderColumn dataField='city' dataSort={ true }>City</TableHeaderColumn>
                    <TableHeaderColumn dataField='contact_name' dataSort={ true }>Contact Name</TableHeaderColumn>
                    <TableHeaderColumn dataField='phone' dataSort={ true }>Phone Number</TableHeaderColumn>
                </BootstrapTable>
            </div>
        )
    }
}
`

And here is the code for the ExpandComponent:
`
class OrganizationDetail extends Component {

    // Called only once when the component is first loaded
    constructor(props) {
        super(props);
        // inherit properties from parent and initialise state of this component
        // we need to use state because state is changeable, while props is passing from parent and not changeable
        this.state = props.organization;
        this.state.errors = {
            organization_name: '', contact_id: '', address: '', city: '', post_code: '', phone: ''
        }

    }

    updateOrganization(event) {
        event.preventDefault();
        const organization = this.props.organization._id;
        const {organization_name, address, post_code, city, phone, contact_id} = this.state;

        // Validation
        var validateContext = Organizations.schema.namedContext("updateOrganization");
        validateContext.validate({organization_name, address, post_code, city, phone, contact_id});
        var invalidKeys = validateContext.invalidKeys();
        invalidKeys = _.map(invalidKeys, function (o) { // get error messages
            return _.extend({message: validateContext.keyErrorMessage(o.name)}, o);
        });

        if (validateContext.isValid()) {
            // call the server side to update this organization data, we need to pass the data for updating
            Meteor.call('organizations.update', organization, this.state, (error)=> {
                if (error) {
                    // console.log(error)
                } else {

                    // should close the expansion (not working) and change row color to normal (working)
                    this.props.unSelectOrganization(organization);
                }
            });
        } else {  // else -> return error msg
            this.showErrorMessages(invalidKeys)
        }
    }

    showErrorMessages(invalidKeys) {
        var errors = {};
        _.map(invalidKeys, (error) => { // mapping errors for each field
            errors[error.name] = error.message
        });
        // console.log(errors);
        this.setState({...this.state, errors})
    }

    handleContactChange(contact_id) {
        var errors = {...this.state.errors, contact_id:''};
        let contact = Meteor.users.findOne({_id: contact_id});
        this.setState({
            contact_id,
            contact_name: contact ? contact.username : null,
            errors
        });
    }
    // handle the change of content of respective <input> caller
    // name: is used to refer back which <input> value need to be updated
    // event: is used to track the change in value
    handleChange(name, event) {
        var change = {};
        var errors = {...this.state.errors, [name]:''};
        change[name] = event.target.value;
        change['errors'] = errors;
        // console.log(change);
        this.setState(change);
    }

    render() {

        const {organization_name, address, post_code, city, phone, contact_id, errors} = this.state;
        // console.log(this.state);

        return (
            <form className="form-inline">
                <div className="col-md-10">
                    <Textfield
                        widthClass="col-md-3"
                        label="Organization Name"
                        labelPlace="top"
                        value={organization_name}
                        error={errors.organization_name}
                        onChange={this.handleChange.bind(this, 'organization_name')}
                    />

                    <Textfield
                        widthClass="col-md-3"
                        label="Address"
                        labelPlace="top"
                        value={address}
                        error={errors.address}
                        onChange={this.handleChange.bind(this, 'address')}
                    />

                    <Textfield
                        widthClass="col-md-3"
                        label="Postal Code"
                        labelPlace="top"
                        value={post_code}
                        error={errors.post_code}
                        onChange={this.handleChange.bind(this, 'post_code')}
                    />

                    <Textfield
                        widthClass="col-md-3"
                        label="City"
                        labelPlace="top"
                        value={city}
                        error={errors.city}
                        onChange={this.handleChange.bind(this, 'city')}
                    />

                    <Textfield
                        widthClass="col-md-3"
                        label="Phone"
                        labelPlace="top"
                        placeholder="phone"
                        value={phone}
                        error={errors.phone}
                        onChange={this.handleChange.bind(this, 'phone')}
                    />

                    <SearchContacts
                        widthClass="col-md-3"
                        label="Contact Name"
                        labelPlace="top"
                        value={contact_id}
                        error={errors.contact_id}
                        handleSelectChange={this.handleContactChange.bind(this)}
                    />
                </div>

                <div className="col-md-2 align-center">
                    <div className="btn-trash">
                        <i className="glyphicon glyphicon-trash"
                           onClick={this.props.deleteOrganizations.bind(this)}/>
                    </div>

                    <button
                        onClick={this.updateOrganization.bind(this)} 
                        className="btn btn-raised btn-primary">
                        Save
                    </button>
                </div>
            </form>
        )
    }
}

`

Do you have any suggestion on this?
Thank you so much for your time and patience.

**Comment by AllenFang** [comment]
I'll try to reproduce this issue without redux just with a local state, but before I start it, I wanna check one thing, is `this.props.selectedOrganizations` always like your expectation in the render method of `OrganizationsTable`, for example, after you click some row or click the button, `this.props.selectedOrganizations` is always correct?

**Comment by ZhuQinglei** [comment]
@AllenFang Actually not always, The first time I click on a row,`this.props.selectedOrganizations` will give me repeated ids like `this.props.selectedOrganizations = ['id1', id1"]`, but it only happens on the 1st click. After that, it is working correct. I am still trying to locate the reason for the first click issue.

**Comment by ZhuQinglei** [comment]
@AllenFang I just created an simple repo with redux implemented here: [ManagingExpansion](https://github.com/ZhuQinglei/ManagingExpansion),  hope it can help with resolving the issue.

**Comment by AllenFang** [comment]
@ZhuQinglei, ok, I'll check this out, but if it was a bug, I may not fix immediately, I'm busy on my work, sorry

**Comment by ZhuQinglei** [comment]
@AllenFang Have you tried the min repo? This is my first time to build a github repo so if there is any error or inconvenience pls tell me and I can fix it quickly.

**Comment by ZhuQinglei** [comment]
@AllenFang  After closer look at this issue, I actually fixed it for my project by manually hide/show the expandable component using jQuery. 

I scrutinised the behaviour of reactDOM element on row click and found the cause of the bug.

The bug is due to that when I click save button in the expandable component, the selecting is working properly to unselect the row (the checkbox is unchecked, and row color changed back to normal) but the expanding is not working (not collapsed, I believe) due to some reason. 

I looked at the html and found the table row that expected to collapse do not have the hidden property while other expandable component table row has, and therefore it is not collapsed. If I select the same row again, the row will be selected while the expansion collapse (doing the opposite) so I guess there is some logic missing to control to expansion together with selection.

There is a manual fix if anyone is facing the same issue: 
use jQuery to override the hide/show of the table row for the expandable component.

Here is my code:
1. in your customised row select function: 
```
handleRowSelect(row, isSelected, e) {
        console.log(e.target);  
        if (isSelected) {
            this.props.selectOrganization(row._id);
            $(e.target).closest('tr').next('tr').show();  // <--- get the next tr of the clicked row, which is the expandable component, show() will display the component
        } else {
            this.props.unSelectOrganization(row._id);
            $(e.target).closest('tr').next('tr').hide(); // <--- hide the component
        }
        return false;
    }
```

2. in your expandable component, hide the table row on click save button, here is the save button function:
```
 updateOrganization(event) {
        event.preventDefault();
        // do Validation
        ..........
        if (validateContext.isValid()) { // update if valid
            Meteor.call('organizations.update', organization, this.state, (error)=> {
                if (error) {
                    Alert.error(error.reason); // alert error
                } else {
                    Alert.success(alertMessages.organization.updateOne);
                    // close the expansion and change row color to normal

                    /* jQuery to close, need to check with the author */
                    //  get the reactDOM and find the table row for this expandable component, and hide it
                    $(ReactDOM.findDOMNode(this)).closest('tr').hide();  // <--- this line
                    unSelectOrganization(organization);
                }
            });
        } else {  // else -> return error msg
            this.showErrorMessages(invalidKeys)
        }
    }
```

And thats all we need to do. 
![screen shot 2017-03-17 at 2 01 36 pm](https://cloud.githubusercontent.com/assets/16860302/24031208/4107073c-0b1c-11e7-93b4-5037914b9082.png)

**Comment by AllenFang** [comment]
I'll review this issue again after my business trip, sorry for long fixing..

**Comment by ZhuQinglei** [comment]
@AllenFang Thank you so much for all the help, if there is anything I can contribute (any issue with mini repo [ManagingExpansion](https://github.com/ZhuQinglei/ManagingExpansion), please tell me and I will fix it asap.  :)

**Comment by AllenFang** [comment]
@ZhuQinglei, nice repo, I'll use your repo to reproduce this issue, thanks a lots

**Comment by AllenFang** [comment]
ah.....  your issue is that `options.expanding` is not work when `this.props.selectedOrganizations` changed?  if yes, this issue was fixed on `v3.1.7`([6f0b5fb](https://github.com/AllenFang/react-bootstrap-table/commit/6f0b5fbe81935b2f4dcf8c709f2f32730a934241))

**Comment by ZhuQinglei** [comment]
@AllenFang Thank you so much for the great work! it's working fine now! you saved my life!! Gonna close this issue now !
